ecs/daemon/outputs/cloudwatch_logs/task.conf

```diff
@@ -0,0 +1,30 @@
+
+[FILTER]
+    Name              ecs
+    Match             task.var.lib.docker.containers.*
+    ECS_Tag_Prefix    task.var.lib.docker.containers.
+    ADD               container    $ECSContainerName
+    ADD               ecs_task_id  $TaskID
+    ADD               ecs_family   $TaskDefinitionFamily
+
+[FILTER]
+    Name               aws
+    Match              task.var.lib.docker.containers.*
+    imds_version       v1
+    az                 false
+    ec2_instance_id    true
+
+
+[OUTPUT]
+    Name                   cloudwatch_logs
+    Match                  task.var.lib.docker.containers.*
+    # log group/stream names with metadata templating
+    log_group_template     /aws/ecs/${CLUSTER_NAME}/application/$ecs_family
+    log_stream_template    $ec2_instance_id.$ecs_task_id.$container
+    # fallback group if templating fails
+    log_group_name         /aws/ecs/${CLUSTER_NAME}/application/fallback-group-for-failed-metadata-templating
+    # fallback stream to log_stream_prefix + tag (log file name) if templating fails
+    log_stream_prefix      json-log-driver-file-
+    region                 ${REGION_TASK}
```

Comment:
bingfeng: So this enables cross region log, but what if customers want cross account log? 

wesley: That would require a different output file.

Reply:
cross account requires you to add a new parameter in teh config:

```
log_stream_prefix      json-log-driver-file-
region                 ${REGION_TASK}
role_arn              ${ROLE_ARN_TASK}
```

which then can't be empty or it will fail validation.

Reply:
we can use the EKS helm chart as an inspiration:

```
log_stream_prefix      json-log-driver-file-
region                 ${REGION_TASK}
${EXTRA_CLOUDWATCH_TASK}
```

its safe for `EXTRA_CLOUDWATCH_TASK`

Need to check if this really works. 

set `EXTRA_CLOUDWATCH_TASK=`role_arn   my-role`

Reply:
need to think about the experience

Reply:
```
log_stream_prefix      json-log-driver-file-
region                 ${REGION_TASK}
${OPTIONAL_ROLE_ARN}
${OPTIONAL_LOG_KEY}
${OPTIONAL_LOG_RETENTION}
```

all of these **I THINK** but need to check, can be empty and it won't break.

Reply:
If this doesn't work, then may be need an init feature which is like what the helm chart.

This might add a newline each time if its empty which won't work.

Reply:
need to check.

Reply:
basically init templating of config

Reply:
## Templating Documentation
One option discussed in today's meeting which I can put here for documentation is as follows:
Have a single environment variable called "init-config" which can be set to a JSON string. JSON is great because any JSON can be reduced to a one line string, thus one environment variable.


May take some more thinking to settle on a good JSON schema but the following would work
#### Configuration JSON
```
{
    /* These are global configuration key value pairs */
    "config": {
        "param1": "val"
    },
    "options": [
         /* Here are a list of options that the customer can configure separately. Name describes the option they are opting into, config is the option's setting */
         {
            "name": "host",
            /* Each config has a default setting, so config is completely optional */
            "config": {
                "destinations": ["s3", "cloudwatch"]
            }
         },
         {
            "name": "host",
            "config": {
                "destination": "cloudwatch",
                "region": "auto" /* Each config gets translated into a template model which gets translated to the template. Non trivial config options should be considered */
            }
         }
    ]
}
```


### Fluent Bit Config Generators
Instead of having init process stitch together raw config files, we introduce a concept of data models and config file templates.

A process on startup parses the JSON config and generates a set of configuration files which are used to startup fluent bit (similar to the init process but with an added step of generating the config files from template)

**Templates**
We can use a widely used templating engine to generate our fluent bit config files adaptively from our config file.
I recommend Mustache render https://mustache.github.io/
It's great, super flexible and allows for conditional and looped blocks of template content which will no doubt come in handy for some future features. See Demo: https://mustache.github.io/#demo

**Data Models**
A mustache template needs a data model to render into it's result. Datamodels are just object. We can generate a Data Model for each template.

**Generator Structure**
This could be how we store the template rendering logic and organize our generator scripts
```
{
    "options": [
        /* template providers for the options we can enable */
        {
            "name": "<OptionName>" /* this option name corresponds to the options that can be enabled */
            "synthesizeDataModel": function(globalConfig, optionConfig) => someObject, /* a function that makes our template data model */
            "templates": [
                /* A list of templates that get processed with our data model and sent to the output destination conf file */
                {
                    "path": "/anyPathToOurTemplate",
                    "destination": "someDestinationConfigFile" /* append to destination file, maybe /parser.conf or /fluent-bit.conf */
                },
               ...
            ]
        }
    ]
}
```

(I highly recommend Typescript, because I love TS and could code this efficiently, but adding node to our image could introduce security scan vulnerabilities that we may have to patch which is unfortunate)

Note: I actually don't like the proposed schemas and structures. It's just a general idea of some thoughts (initial thoughts).

Reply:
Just having the log_key option, either via this or just another config file, for the task logs, is something I kind of want to have even just for the first launch 🤔

Reply:
I think doing the `log_key log` thing is what a lot of users want for their container logs. They don't want them as JSONs.